### ISSUE TITLE
Pre-golang 1.6 Dockers can't talk to golang 1.6 Docker

### ISSUE THREAD

**Issue by duglin** [docker/docker#20865]
When Docker v1.9 (and before) talks to Docker v1.11 (master) via a unix socket the daemon will reject the request with this error:
```
Error response from daemon: 400 Bad Request: malformed Host header
```

I believe this is because the HTTP Host header will have a value of: `/var/run/docker.sock` which isn't a valid hostname and I believe golang 1.6 is now checking the value in validHostHeader() (in https://golang.org/src/net/http/request.go) and "/" isn't a valid char.

In Docker v1.11 (master) we now set the HTTP Host header to "", which is why we don't see this on master.

I've been looking into a way to avoid having golang perform this check but haven't found a way yet.

Net is that this may force all clients that want to talk to a daemon via a unix docker to be upgraded at the same time as the daemon. Not an ideal UX situation.

The only other option I can think of, and its not great, is to postpone upgrading to golang 1.6 until v1.12 so that we can make v1.11 send a Host header with a value of "" for one release before we turn on this check in the daemon.  Then during the v1.12 upgrade process clients can upgrade later then the daemon, as long as they're at least on v1.11.

Anyway, wanted to open an issue so we could have a discussion and brainstorm on ideas.

**Comment by dqminh** [comment]
I dont think the actual impact is that great. In which case where unix socket is involved do you not want to upgrade the cli / daemon together ? The only case i think of is that a cli inside container talks to a daemon in the host.

Does patching  v1.10 so it sends correct host header a viable alternative ?

**Comment by duglin** [comment]
@dqminh yea I have no idea how big of an impact it is but I figured I'd start out with a high priority to ensure its not missed.  Personally, this breaks my build env because I tend to use old Dockers on the CLI when I build.

**Comment by duglin** [comment]
Updated description - this is just an issue for docker clients that are v1.9 and older.  Docker client v1.10 doesn't have this issue.

**Comment by WeiZhang555** [comment]
I don't think the impact is that great either. If CLI is talking to Daemon via unix socket, that means they are in the same host, right? Why don't update CLI to same version with daemon?

**Comment by cpuguy83** [comment]
ping @docker/core-engine-maintainers 

Are we ok with this behavior?
I agree it's weird that someone would have a docker client going over the unix socket that is a different binary from the daemon....

TLDR; engine built with go 1.6 can't talk to clients built with < go 1.5 over a unix socket due to stricter enforcement of the HTTP spec in go.

**Comment by jfrazelle** [comment]
i think its a problem no?

**Comment by ibuildthecloud** [comment]
This is a huge problem.  You can't ignore the fact that a lot of containers bind mount `/var/run/docker.sock` and package their own Docker client.  When you do this you are force to package the lowest common denominator which for me has been Docker 1.6.  Please fix.  This really breaks a lot of my stuff and is currently a blocker for me to more thoroughly test Docker 1.11.

**Comment by tiborvass** [comment]
Just to make sure people realize that the issue seems to be in Go 1.5+, so it would require us to downgrade to Go 1.4.

**Comment by tiborvass** [comment]
Ok so after testing this, this is the situation:

```
client    ->  daemon
------------------------------------
1.9       ->   1.10      OK
1.9       ->   1.11    **KO**
1.10      ->   1.11      OK
```

**Comment by jfrazelle** [comment]
but 1.10 was built with 1.5, so we shouldnt have to downgrade to 1.4 right?

**Comment by jfrazelle** [comment]
:confused:

**Comment by cpuguy83** [comment]
Yeah, 1.5 looks good to me.

**Comment by duglin** [comment]
For more chatter... https://botbot.me/freenode/docker-dev/2016-03-23/?msg=62715258&page=1

**Comment by ibuildthecloud** [comment]
Okay, so probably the sanest approach is just go with golang 1.5 and hope the future holds a better solution.  Otherwise I do have a functioning proof of concept on how we can fix this: https://github.com/docker/docker/pull/21423

**Comment by jhowardmsft** [comment]
Hmmm. This now introduces an interesting cross platform issue though. There is no reason to not use 1.6 on Windows. There is no legacy support issue until RTM. I also believe containerd censoring scheme as relies on being built with 1.6. But the official Windows binary is cross built from Linux, not on Windows. So that means the official build won't be what is validated on the Windows. I really don't want to drop back on Windows unless the cross-binary is built with 1.6. @icecrime @mikedougherty

**Comment by calavera** [comment]
`/var/run/docker.sock` is an invalid Host value. We only got away with it because Go didn't validate the Host header until Go 1.6, which was a bug in their HTTP 1.1 implementation:

https://github.com/golang/go/issues/13624

Sorry, but I don't think we should preserve bugs as valid behaviors.

**Comment by ibuildthecloud** [comment]
@calavera I think that is a bit too idealistic.  I already know this is a massive impact to me to not being able to use old Docker clients.  As I mentioned before I haven't been able to fully test 1.11 because of this issue.  It requires me to rebuild far too many of my images, systems, etc.  Docker 1.9 is not that old.  If we do not support older clients this will be the first release of Docker that I will have to tell my users that we can't support on day one.  And when I do support 1.11 I will mostly likely have to drop support for Docker <=1.9, which again I don't want to do.  If we are expecting people to rebuild their entire infrastructure on Docker, we must ensure solid compatibility which often means we sacrifice our software engineering ideologies.

If golang 1.6 is important I am more that happy to provide a robust version of #21423 to make this work.

**Comment by calavera** [comment]
> If golang 1.6 is important

It's not important, at least for this version. but we should not keep bad behaviors around.

**Comment by ibuildthecloud** [comment]
@calavera This strikes a chord with me, sorry.  A general concern I have with Docker is that there is far too much concern for the "right" or "wrong" way to build software and not enough concern for stable and supportable.  Maybe I'm just getting old.

**Comment by duglin** [comment]
> Maybe I'm just getting old.

I'll save a rocking chair and some pink lemonade for ya...  :-)

**Comment by calavera** [comment]
> not enough concern for stable and supportable

From my point of view, ignoring something as basic as the HTTP 1.1 spec is not very supportable and stable. The right solution would be to release patches with the right host for versions we want to support, but that's not going to happen either.

**Comment by ibuildthecloud** [comment]
Yikes.  Are we really going to just force users to upgrade?  I see RC1 is cut and is still broken.

**Comment by bboreham** [comment]
Weave Net ships a cope of the Docker CLI inside a container so we can call it from a script, so this issue gives us a choice between an image that works for Docker daemon versions 1.6-1.10, or one that works for Docker 1.10 and 1.11 only. Which is not a great user experience.

**Comment by duglin** [comment]
At first I was thinking that this wasn't that big of an issue, but after the feedback from the community I'm now leaning towards the belief that we do need to offer some kind of solution - other than "just upgrade".

One option is to pick-up @ibuildthecloud's solution ( #21423 ) but make it available via a temporary daemon-side flag/env-var - since its not a lot of code (even after we clean it a little ;-) ) and it can be isolated to be easily removed later.

It wouldn't take a lot to convince me that we should turn it on by default though, at least for v1.11.

**Comment by cpuguy83** [comment]
I don't see a reason to have such a hack for one version or even to revert to 1.5 for just 1 version. It'd have to be that way until we drop support for these older clients.

**Comment by bboreham** [comment]
Mapping the Unix socket into a container is reasonably popular (5,523 matches):

https://github.com/search?q=%22-v+%2Fvar%2Frun%2Fdocker.sock%22&type=Code&utf8=%E2%9C%93

There must be a lot of those containers in the wild with older implementations.

**Comment by icecrime** [comment]
@bboreham That's not quite true: you can ship a single image with a modified client that happens to _not_ send an invalid `Host:` header to the daemon.

**Comment by crosbymichael** [comment]
Also a better model for mapping the client and socket into the container is to bind mount both so you have exactly the same version outside as inside the container instead of baking the docker binaries into your containers.  

This is the correct way to do it:

`docker run -v /run/docker.sock:/run/docker.sock -v /usr/local/bin/docker:/usr/local/bin/docker ...`

Anyways we will revert back to go 1.5.3 because of this issue and look at upgrading in the future.

**Comment by ibuildthecloud** [comment]
@crosbymichael But i have `docker` installed at `/usr/bin/docker` :smile: 

Thanks for addressing this

**Comment by thaJeztah** [comment]
@ibuildthecloud then you're doing it wrong! haha

**Comment by rhatdan** [comment]
@crosbymichael solution would only work if you have a statically built docker or the OS of the host is the same as the OS in the container, or if you are very lucky.  Shared libraries can be tricky.

**Comment by cpuguy83** [comment]
I'm going to reopen this since we still will need to be able to upgrade go versions at some point and it's still a problem to deal with.

**Comment by cpuguy83** [comment]
FYI, it looks like golang 1.7 is planned for July/August time-frame so we may not need to really deal with this until 1.13.

**Comment by runcom** [comment]
reopening just to have this around since the related PR is in code review

**Comment by berolinux** [comment]
FWIW for people who need it right now, here's a backport of the patch that works with 1.11.1:
https://github.com/OpenMandrivaAssociation/docker/blob/master/docker-1.11.1-work-with-old-clients.patch

**Comment by thaJeztah** [comment]
@berolinux or direct from the PR that was merged? https://patch-diff.githubusercontent.com/raw/docker/docker/pull/22000.patch. However docker 1.11.1 is compiled with Go 1.5.4, so there's no need to apply that patch

**Comment by berolinux** [comment]
@thaJeztah 22000 doesn't apply on top of 1.11.1 -- had to do some editing to make it fit.
And it's needed for those of us who compile ourselves (pretty much all Linux distros do - and few still ship 1.5).
